Silence healthcheck requests from the log

dhh opened the pull request:
The default Rails health check is intended to be hit every second or so by Kamal to ensure that the application is still or has just become responsive. This clutters up the production logs and makes it hard to follow what's actually going on. This adds a new middleware called `Rails::Rack::SilenceRequest` that is used by a new config `config.silence_healthcheck = path` to prevent that.

dhh committed:
Add SilenceRequest middleware

dhh committed:
Make config.silence_healthcheck a way to quiet "/up"

dhh committed:
The remote ip part is actually not necessary

dhh committed:
No longer used

dhh committed:
Add CHANGELOG

dhh committed:
Make it more explicit that the argument is a path

dhh committed:
Better explain the purpose

dhh committed:
No need to repeat the env

dhh committed:
Why oh why

skipkayhil commented on railties/lib/rails/rack/silence_request.rb:
```diff
@@ -0,0 +1,37 @@
+      def call(env)
+        request = ActionDispatch::Request.new(env)
+
+        if request.path == @path
```
```suggestion
        if request.path === @paths
```

What would you think about making it a configurable array of paths (String/Regex)? Then we could do something like:

```
config.silent_request_paths << "/assets"
```

in Sprockets/Propshaft instead of having the QuietAssets middleware be almost a copy of this one.

dhh replied:
I had that at first, but I then arrested myself with a charge of YAGNI. I couldn't come up with any other use cases except for the healthcheck, so decided to focus on that. The API is a lot more straight forward than something generic.

dhh replied:
But I still think we could reuse this. We could inherit the QuietAssets middleware in Propshaft from this and just hardset the path:. But not sure it's even worth it. The implementation is so small here.

dhh replied:
In fact, I just made the implementation even slimmer by borrowing from Propshaft's quiet asset middleware. So thanks for the reminder of that one!

Earlopain replied:
I use ActiveStorage with the disk controller in dev and the amount of logs I get from that are astronomical, so it would be appreciated if I was able to configure this middleware for that.

I see that it is very slim so I'll probably just joink it for myself if needed but I do believe there is a need here: https://stackoverflow.com/questions/52147873/mute-activestorage-logs

dhh committed:
Document the new setting

dhh committed:
Slim down the implementation

Thanks to @skipkayhil for reminding me of Propshaft::QuietAssets which
does something very similar and does it just via env!

skipkayhil commented on guides/source/configuring.md:
```diff
@@ -593,6 +593,14 @@ config.session_store :my_custom_store
 
 The default store is a cookie store with the application name as the session key.
 
+#### config.silence_healthcheck_path
```
```suggestion
#### `config.silence_healthcheck_path`
```

I think this should fix the Lint check in CI, although that error message could definitely be more friendly 🤔

dhh committed:
Update guides/source/configuring.md

Co-authored-by: Hartley McGuire <skipkayhil@gmail.com>

dhh committed:
Double quotes

ghiculescu commented on railties/test/commands/middleware_test.rb:
```diff
@@ -196,10 +196,10 @@ def app
     assert_includes middleware, "ActionDispatch::AssumeSSL"
   end
 
-  test "ActionDispatch::SSL is present when force_ssl is set" do
```
@dhh did you mean to delete this test?

dhh replied:
I did not. Not sure what happened there. Please do PR to restore!

ghiculescu replied:
https://github.com/rails/rails/pull/52812

AnalyzePlatypus commented:
Will this ship in Rails 8, or in an earlier point release?

dhh commented:
8. We don't back port features.

cristianbica commented:
@dhh how do you feel abstracting this to a `config.silence_requests_paths = ['/up', '/others']`?
I'm thinking at use-cases like an endpoint exporting some metrics, people running rack-mini-profiler in production or people just filtering out logging for some endpoints which just adds noise to logging.

dhh commented:
@cristianbica Are any of those examples you mention problems you've actually hit in production? At what intervals are those requests running? I don't want to blanket rule out making this more generic, but I find that you're better off with something specific, unless you've proven without a doubt the need for something generic.

adampope commented:
@dhh In our app we have an async request for red dot / badge counts that happens on most page loads.  I recently wrote a very similar middleware to filter those out of the development logs as I got bored of scrolling back up the tail past the badge request to see details of the main request.  Having something configurable like @cristianbica is suggesting, and being able to be different in different environments would certainly have been great for our real use case.

cristianbica commented:
> @cristianbica Are any of those examples you mention problems you've actually hit in production? At what intervals are those requests running? I don't want to blanket rule out making this more generic, but I find that you're better off with something specific, unless you've proven without a doubt the need for something generic.

@dhh on my current projects I do not have a valid use-case to prove the need to generalize this functionality.

In the past projects I've implemented different solution to reduce logging (e.g. silencer gem for requests). The purpose was to reduce clutter in the logs and reduce costs. From what I remember:
- we had an endpoint exposing sidekiq metrics called minutely
- we had many webhook-like requests - those were recorded and processed in the background; logging was too verbose so there was no benefit to log them

Anyway, looking at the implementation in this PR, one could easy filter out logging for other paths by adding to `application.rb` multiple lines like this

```ruby
config.middleware.insert_before ::Rails::Rack::Logger, :Rails::Rack::SilenceRequest, path: '/sidekiq_stats'
```

morgoth commented:
@dhh With thruster gem included by default, the "/up" requests are showing up in the logs anyway. Is there any way to silence that as well, as I don't see such possibility in thruster project yet?
